tutorial/docker-compose-vitess.yaml

```diff
@@ -0,0 +1,37 @@
+version: '2'
+services:
+  zookeeper:
+    image: debezium/zookeeper:${DEBEZIUM_VERSION}
+    ports:
+      - 2181:2181
+      - 2888:2888
+      - 3888:3888
+  kafka:
+    image: debezium/kafka:${DEBEZIUM_VERSION}
+    ports:
+      - 9092:9092
+    links:
+      - zookeeper
+    environment:
+      - ZOOKEEPER_CONNECT=zookeeper:2181
+  vitess:
```

Comment:
How about making this an actual cluster with say three nodes, so people can explore the sharding semantics of Vitess?

Reply:
This vitess container actually already has multiple shards inside. Inside the container, there are:

- 1 sharded keyspace that has 2 shards, each shard is a MySQL cluster that has 3 processes (1 master, 1 replica and 1 slave)
- 1 unsharded keyspace that is itself a MySQL cluster that has 3 processes (1 master, 1 replica and 1 slave).

So in total, we have 9 MySQL processes (6 for the sharded keyspace, 3 for the unsharded keyspace) inside the container.

Reply:
Oh, I see, that's interesting. I guess there's no way then though to e.g. take down a shard? OTOH, I suppose running nine separate containers might be a tad too much 🤔 .

Reply:
> I suppose running nine separate containers might be a tad too much.

Yeah, running nine processes within a container might be a bit too much. Here's the resource (CPU/MEM) utilization of the containers on my laptop:
```bash
>docker stats
CONTAINER ID   NAME                   CPU %     MEM USAGE / LIMIT     MEM %     NET I/O         BLOCK I/O         PIDS
f33eb73c9013   tutorial_connect_1     2.54%     803.5MiB / 19.56GiB   4.01%     150kB / 118kB   85MB / 106kB      42
7ab005a3059a   tutorial_kafka_1       3.24%     396MiB / 19.56GiB     1.98%     184kB / 221kB   19.6MB / 492kB    73
c00482db3ae8   tutorial_zookeeper_1   0.29%     84.45MiB / 19.56GiB   0.42%     73.1kB / 65kB   49.2MB / 537kB    52
4d54a59ae195   tutorial_vitess_1      45.26%    2.327GiB / 19.56GiB   11.90%    1.33kB / 0B     90.6MB / 1.69GB   727

```

The [official Vitess local docker setup](https://github.com/vitessio/vitess/tree/master/docker/local) also uses multiple processes within one container.

I think the tricky part is that **sharded keyspace**(s) store their sequences in an **unsharded keyspace**. Each shard is itself a MySQL cluster (one master, one replica slave, and one rdonly slave). The vitess docker image used by the Vitess Connector's integration tests also has this setup (9 processes within one container) to have sharded keyspace available for the test.

A few optimized options:

Option 1:
If it's too much, I can reduce the MySQL cluster to have only one master process, therefore, we'd only have in total 3 processes within the container. However, Vitess Connector then has to subscribe to the MySQL master instance, which is not usually the case in a production environment. In production, Vitess Connector usually subscribes to either replica or rdonly slave.

Option 2:
I can further reduce the resource utilization by having only one unsharded space in the example tutorial, which consists of a MySQL cluster of only one master. Therefore, only 1 process running.

Option 3:
Use 9 services in the docker-compose? 3 services for the unsharded keyspace (master, replica, rdonly) and 6 services for the sharded keyspace (master/replica/rdonly for shard No.1, master/replica/rdonly for shard No.2)?

WDYT?

Reply:
>I guess there's no way then though to e.g. take down a shard?

Indeed, if we keep all the processes in the same container (a.k.a docker-compose service), we'll need to `docker exec -it tutorial_vitess_1 bash` into the container to kill all the 3 MySQL processes (master, replica and rdonly) of a given shard.

Reply:
I draw an architecture diagram to help understand the number of processes in the Vitess container. Actually, each square and cylinder represents a process :)
![vitess_container](https://user-images.githubusercontent.com/1978545/106108913-9a114380-6151-11eb-8cfb-8af49e1858d9.png)

Reply:
Wow, that's really helpful. Thanks a lot, @keweishang!